File: EBBannerView/Classes/EBBannerWindow.m
//
//  EBBannerWindow.m
//  demo
//
//  Created by pikacode@qq.com on 2017/10/23.
//  Copyright © 2017年 pikacode@qq.com. All rights reserved.
//

#import "EBBannerWindow.h"
#import "EBBannerViewController.h"
#import "EBBannerView+Categories.h"

@implementation EBBannerWindow

static EBBannerWindow *sharedWindow;

+(instancetype)sharedWindow{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        SEL sel = @selector(initWithWindowScene:);
        if ([EBBannerWindow.new respondsToSelector:sel]) {
            id obj = [UIApplication.sharedApplication.keyWindow valueForKey:@"windowScene"];
            #pragma clang diagnostic push
            #pragma clang diagnostic ignored "-Warc-performSelector-leaks"
            sharedWindow = [[EBBannerWindow alloc] performSelector:sel withObject:obj];
            #pragma clang diagnostic pop
        } else {
            // Fallback on earlier versions
            sharedWindow = [[self alloc] initWithFrame:CGRectZero];
        }
        sharedWindow.windowLevel = UIWindowLevelNormal;
        sharedWindow.layer.masksToBounds = NO;
        [sharedWindow setHidden:NO];
        
        [EBBannerViewController setSupportedInterfaceOrientations:UIInterfaceOrientationMaskPortrait | UIInterfaceOrientationMaskLandscape];
        [EBBannerViewController setStatusBarHidden:NO];
        
        EBBannerViewController *vc = [EBBannerViewController new];
        vc.view.backgroundColor = [UIColor clearColor];
        vc.view.frame = CGRectMake(0, 0, ScreenWidth, ScreenHeight);
        sharedWindow.rootViewController = vc;
    });
    return sharedWindow;
}

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        [self addObserver:self forKeyPath:@"frame" options:NSKeyValueObservingOptionNew|NSKeyValueObservingOptionOld context:nil];
    }
    return self;
}

-(UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event{
    __block UIView *view;
    [self.rootViewController.view.subviews enumerateObjectsUsingBlock:^(__kindof UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (CGRectContainsPoint(obj.frame, point)) {
            view = obj;
        }
    }];
    if (view) {
        CGPoint point1 = [self convertPoint:point toView:view];
        return [view hitTest:point1 withEvent:event];
    } else {
        if (@available(iOS 13.0, *)) {
            if (UIApplication.sharedApplication.windows.count > 0) {
                UIWindow *window = UIApplication.sharedApplication.windows[0];
                if (window.isKeyWindow) {
                    return [window hitTest:point withEvent:event];
                }
            }
            //iOS13以后，keyWindow不再是最开始创建的window，而是当前显示的window，这么写会造成某些场景死循环。
//            return [UIApplication.sharedApplication.keyWindow hitTest:point withEvent:event];
        }
        return [super hitTest:point withEvent:event];
    }
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context{
    if ([keyPath isEqualToString:@"frame"] && !CGRectEqualToRect(self.frame, CGRectZero)) {
        self.frame = CGRectZero;
    }
}

-(void)dealloc{
    if (@available(iOS 13.0, *)) {
        
    } else {
        [self removeObserver:self forKeyPath:@"frame"];
    }
}

@end

File: EBBannerView/SwiftClasses/EBBannerWindow.swift
//
//  EBBannerWindow.swift
//  EBBannerViewSwift
//
//  Created by pikacode on 2020/1/2.
//

import UIKit

class EBBannerWindow: UIWindow {

    @available(iOS 13.0, *)
    public override init(windowScene: UIWindowScene){
        super.init(windowScene: windowScene)
    }
    
    static let shared: EBBannerWindow = {
        var window: EBBannerWindow
        if #available(iOS 13.0, *) {
            window = EBBannerWindow(windowScene: UIApplication.shared.keyWindow!.windowScene!)
        } else {
            window = EBBannerWindow(frame: .zero)
        }
        window.windowLevel = .alert
        window.layer.masksToBounds = false
        let keyWindow = UIApplication.shared.keyWindow
        window.makeKeyAndVisible()
        /* fix bug:
         EBBannerViewController setSupportedInterfaceOrientations -> Portrait
         push to a VC with orientation Left
         UITextFiled's pad will show a wrong orientation with Portrait
         */
        EBEmptyWindow.shared.makeKeyAndVisible()
        keyWindow?.makeKeyAndVisible()
        EBBannerController.setSupportedInterfaceOrientations(value: [.portrait, .landscape])
        EBBannerController.setStatusBarHidden(hidden: false)
        let vc = EBBannerController()
        vc.view.backgroundColor = .clear
        let size = UIScreen.main.bounds.size
        vc.view.frame = CGRect(x: 0, y: 0, width: size.width, height: size.height)
        window.rootViewController = vc
        return window
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addObserver(self, forKeyPath: "frame", options: [.new, .old], context: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        if #available(iOS 13.0, *) {
            print("")
        } else {
            removeObserver(self, forKeyPath: "frame")
        }
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "frame" && !frame.equalTo(.zero) {
            frame = .zero
        }
    }
    
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        var view: UIView?
        for v in rootViewController?.view.subviews ?? [UIView]() {
            if v.frame.contains(point) {
                view = v
                break
            }
        }
        if view == nil {
            if #available(iOS 13.0, *) {
                return UIApplication.shared.keyWindow?.hitTest(point, with: event)
            } else {
                return super.hitTest(point, with: event)
            }
        } else {
            let p = convert(point, to: view)
            return view?.hitTest(p, with: event)
        }
    }
    
}
  
 
